src/daft-writers/src/lib.rs

```diff
@@ -151,43 +168,65 @@ pub fn make_catalog_writer_factory(
     }
 }
 
-fn calculate_target_parquet_rows(
-    estimated_row_size_bytes: f64,
-    target_filesize: f64,
-    target_row_group_size: f64,
-    inflation_factor: f64,
-) -> (usize, usize) {
-    let target_in_memory_file_size = target_filesize * inflation_factor;
-    let target_in_memory_row_group_size = target_row_group_size * inflation_factor;
-
-    let target_file_rows = if estimated_row_size_bytes > 0.0 {
-        target_in_memory_file_size / estimated_row_size_bytes
-    } else {
-        target_in_memory_file_size
-    } as usize;
+/// This struct is used to adaptively calculate the target in-memory size of a table
+/// given a target on-disk size, and the actual on-disk size of the written data.
+pub(crate) struct TargetInMemorySizeBytesCalculator {
+    estimated_inflation_factor: AtomicU64, // Using u64 to store f64 bits
+    target_size_bytes: usize,
+    num_samples: AtomicUsize,
+}
 
-    let target_row_group_rows = min(
-        target_file_rows,
-        if estimated_row_size_bytes > 0.0 {
-            target_in_memory_row_group_size / estimated_row_size_bytes
-        } else {
-            target_in_memory_row_group_size
-        } as usize,
-    );
+impl TargetInMemorySizeBytesCalculator {
+    fn new(target_size_bytes: usize, initial_inflation_factor: f64) -> Self {
+        assert!(target_size_bytes > 0 && initial_inflation_factor > 0.0);
+        Self {
+            estimated_inflation_factor: AtomicU64::new(initial_inflation_factor.to_bits()),
+            target_size_bytes,
+            num_samples: AtomicUsize::new(0),
+        }
+    }
 
-    (target_file_rows, target_row_group_rows)
-}
+    fn calculate_target_in_memory_size_bytes(&self) -> usize {
+        let factor = f64::from_bits(self.estimated_inflation_factor.load(Ordering::Relaxed));
+        (self.target_size_bytes as f64 * factor) as usize
+    }
 
-fn calculate_target_csv_rows(
-    estimated_row_size_bytes: f64,
-    target_filesize: f64,
-    inflation_factor: f64,
-) -> usize {
-    let target_in_memory_file_size = target_filesize * inflation_factor;
+    pub fn record_and_update_inflation_factor(
+        &self,
+        actual_on_disk_size_bytes: usize,
+        estimate_in_memory_size_bytes: usize,
+    ) {
+        // Avoid division by zero - in practice actual_on_disk_size_bytes should never be 0
+        // as we're dealing with real files, but be defensive
+        if actual_on_disk_size_bytes == 0 {
+            return;
+        }
 
-    if estimated_row_size_bytes > 0.0 {
-        (target_in_memory_file_size / estimated_row_size_bytes) as usize
-    } else {
-        target_in_memory_file_size as usize
+        let new_inflation_factor =
+            estimate_in_memory_size_bytes as f64 / actual_on_disk_size_bytes as f64;
+        let new_num_samples = self.num_samples.fetch_add(1, Ordering::Relaxed) + 1;
+
+        let current_factor =
```

Comment:
this is pretty bootleg and will prob cause bugs in the future. especially if folks try to use `current_factor` directly and get some huge number from it.

You are also using 2 atomics which makes the current factor not atomic.

I would just recommend using a `Mutex<f32>` and doing a mul_add on that. You're not gonna be holding on it for long anyways and we're bottle necked by the GIL mutex anyways.

Reply:
`f32` or `f64`? The default inflation factors in execution config come as `f64`

Reply:
either is fine. we don't need the precision of `f64` but it doesn't hurt